### PR TITLE
arcv: Add a separate multilib configuration list

### PR DESCRIPTION
**Pull request by kolerov**
When GCC is built for `riscv*-snps-*` targets, a separate multilib configuration list must be used to make all configurations close to basic ARC-V profiles as much as possible. To achieve that it is necessary to build libraries not only for `march/mabi` combinations, but rather for `march/mabi/mtune` triplets.

Here is a list of ARC-V targets supported by this configuration list:

* RMX-100 mini
* RMX-100 with variations
* RHX with variations
* RPX with variations

Note that RPX (rv64) configurations are built without `mtune` option since it is not implemented yet for RPX.

Here is a list of multilib configurations after building the toolchain:

```
$ riscv64-snps-elf-gcc -print-multi-lib
.;
rv32e/ilp32e/rmx100;@march=rv32e@mabi=ilp32e@mtune=rmx100
rv32i/ilp32/rmx100;@march=rv32i@mabi=ilp32@mtune=rmx100
rv32i/ilp32/rmx500;@march=rv32i@mabi=ilp32@mtune=rmx500
rv32i/ilp32/rhx;@march=rv32i@mabi=ilp32@mtune=rhx
rv32ic_zcb_zcmp_zcmt_zba_zbb_zbs_zicsr/ilp32/rmx100;@march=rv32ic_zcb_zcmp_zcmt_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=rmx100
rv32imc_zcb_zcmp_zcmt_zba_zbb_zbs_zicsr/ilp32/rmx100;@march=rv32imc_zcb_zcmp_zcmt_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=rmx100
rv32imac_zcb_zcmp_zcmt_zba_zbb_zbs_zicsr/ilp32/rmx100;@march=rv32imac_zcb_zcmp_zcmt_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=rmx100
rv32imac_zcb_zcmp_zba_zbb_zbs_zicsr/ilp32/rhx;@march=rv32imac_zcb_zcmp_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=rhx
rv32imafc_zcb_zcmp_zba_zbb_zbs_zicsr/ilp32f/rhx;@march=rv32imafc_zcb_zcmp_zba_zbb_zbs_zicsr@mabi=ilp32f@mtune=rhx
rv32imafd_zca_zcb_zcmp_zba_zbb_zbs_zicsr/ilp32d/rhx;@march=rv32imafd_zca_zcb_zcmp_zba_zbb_zbs_zicsr@mabi=ilp32d@mtune=rhx
rv64imac_zcb_zba_zbb_zbs_zicsr/lp64;@march=rv64imac_zcb_zba_zbb_zbs_zicsr@mabi=lp64
rv64imac_zcb_zba_zbb_zbs_zicsr/lp64/medany;@march=rv64imac_zcb_zba_zbb_zbs_zicsr@mabi=lp64@mcmodel=medany
rv64imafdc_zcb_zba_zbb_zbs_zicsr/lp64d;@march=rv64imafdc_zcb_zba_zbb_zbs_zicsr@mabi=lp64d
rv64imafdc_zcb_zba_zbb_zbs_zicsr/lp64d/medany;@march=rv64imafdc_zcb_zba_zbb_zbs_zicsr@mabi=lp64d@mcmodel=medany
```

### CODE REVIEW

File: gcc/config/riscv/t-arcv [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#                     rmx100mini  plain       rmx100_base.tcf - m                          rmx100_base.tcf                               rmx100_base.tcf + a                            rhx100_base.tcf                           rhx100_base_fpu.tcf - -zfh                 rhx100_base.tcf - -zcd + d                     rpx                                  rpx + fpu
+MULTILIB_OPTIONS    = march=rv32e/march=rv32i/march=rv32ic_zcb_zcmp_zcmt_zba_zbb_zbs_zicsr/march=rv32imc_zcb_zcmp_zcmt_zba_zbb_zbs_zicsr/march=rv32imac_zcb_zcmp_zcmt_zba_zbb_zbs_zicsr/march=rv32imac_zcb_zcmp_zba_zbb_zbs_zicsr/march=rv32imafc_zcb_zcmp_zba_zbb_zbs_zicsr/march=rv32imafd_zca_zcb_zcmp_zba_zbb_zbs_zicsr/march=rv64imac_zcb_zba_zbb_zbs_zicsr/march=rv64imafdc_zcb_zba_zbb_zbs_zicsr
```
Comment (abrodkin) [inline gcc/config/riscv/t-arcv]: Just wondering if there's a way to have one option per line to improve readability. The same question (but to a lesser extent) is applicable to `MULTILIB_DIRNAMES`.

Comment (kolerov) [reply gcc/config/riscv/t-arcv]: Yes, this file is included by Makefile script. I will make it more readable.

**Comment by kolerov** [comment]
I made it more readable and updated a list of configurations:

```
$ riscv64-snps-elf-gcc -print-multi-lib
.;
rv32e/ilp32e/rmx100;@march=rv32e@mabi=ilp32e@mtune=rmx100
rv32e_zicsr/ilp32e/rmx100;@march=rv32e_zicsr@mabi=ilp32e@mtune=rmx100
rv32em_zce_zba_zbb_zbs_zfinx_zdinx_zicsr/ilp32e/rmx100;@march=rv32em_zce_zba_zbb_zbs_zfinx_zdinx_zicsr@mabi=ilp32e@mtune=rmx100
rv32ema_zce_zba_zbb_zbs_zfinx_zdinx_zicsr/ilp32e/rmx100;@march=rv32ema_zce_zba_zbb_zbs_zfinx_zdinx_zicsr@mabi=ilp32e@mtune=rmx100
rv32i_zce_zba_zbb_zbs_zicsr/ilp32/rmx100;@march=rv32i_zce_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=rmx100
rv32i_zce_zba_zbb_zbs_zicsr/ilp32/rmx500;@march=rv32i_zce_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=rmx500
rv32im_zce_zba_zbb_zbs_zicsr/ilp32/rmx100;@march=rv32im_zce_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=rmx100
rv32im_zce_zba_zbb_zbs_zicsr/ilp32/rmx500;@march=rv32im_zce_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=rmx500
rv32ima_zce_zba_zbb_zbs_zicsr/ilp32/rmx100;@march=rv32ima_zce_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=rmx100
rv32ima_zce_zba_zbb_zbs_zicsr/ilp32/rmx500;@march=rv32ima_zce_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=rmx500
rv32ima_zce_zba_zbb_zbs_zfinx_zicsr/ilp32/rmx100;@march=rv32ima_zce_zba_zbb_zbs_zfinx_zicsr@mabi=ilp32@mtune=rmx100
rv32ima_zce_zba_zbb_zbs_zfinx_zicsr/ilp32/rmx500;@march=rv32ima_zce_zba_zbb_zbs_zfinx_zicsr@mabi=ilp32@mtune=rmx500
rv32ima_zce_zba_zbb_zbs_zfinx_zdinx_zicsr/ilp32/rmx100;@march=rv32ima_zce_zba_zbb_zbs_zfinx_zdinx_zicsr@mabi=ilp32@mtune=rmx100
rv32ima_zce_zba_zbb_zbs_zfinx_zdinx_zicsr/ilp32/rmx500;@march=rv32ima_zce_zba_zbb_zbs_zfinx_zdinx_zicsr@mabi=ilp32@mtune=rmx500
rv32imac_zcb_zcmp_zba_zbb_zbs_zicsr/ilp32/rhx;@march=rv32imac_zcb_zcmp_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=rhx
rv32imafc_zcb_zcmp_zba_zbb_zbs_zicsr/ilp32f/rhx;@march=rv32imafc_zcb_zcmp_zba_zbb_zbs_zicsr@mabi=ilp32f@mtune=rhx
rv32imafd_zca_zcb_zcmp_zba_zbb_zbs_zicsr/ilp32d/rhx;@march=rv32imafd_zca_zcb_zcmp_zba_zbb_zbs_zicsr@mabi=ilp32d@mtune=rhx
rv64imac_zcb_zba_zbb_zbs_zicsr/lp64/medany;@march=rv64imac_zcb_zba_zbb_zbs_zicsr@mabi=lp64@mcmodel=medany
rv64imafdc_zcb_zba_zbb_zbs_zicsr/lp64d/medany;@march=rv64imafdc_zcb_zba_zbb_zbs_zicsr@mabi=lp64d@mcmodel=medany
```

Now there are more configurations for RMX-100 mini and RMX-500.

**Comment by abrodkin** [comment]
@kolerov indeed, it looks way more readable now, thanks!

File: gcc/config/riscv/t-arcv [hunk 2]
```diff
@@ -0,0 +1,89 @@
+empty =
+space = $(empty) $(empty)
+
+# Fallaback configurations
```
Comment (abrodkin) [inline gcc/config/riscv/t-arcv:9]: @artemiy-volkov does that look fine by you or you prefer to add something else?

**Comment by kolerov** [comment]
I've updated all configurations. Now this set of optimized configurations is built for `riscv*-snps-*` targets:

* RMX-100 mini (`-mtune=arc-v-rmx-100-series`):

  - `rv32e`
  - `rv32e_zicsr`
  - `rv32em_zce_zba_zbb_zbs_zfinx_zdinx_zicsr`
  - `rv32ema_zce_zba_zbb_zbs_zfinx_zdinx_zicsr`

* RMX-100 (`-mtune=arc-v-rmx-100-series`, without hard FPU):

  - `rv32i_zce_zba_zbb_zbs_zicsr`
  - `rv32im_zce_zba_zbb_zbs_zicsr`
  - `rv32ima_zce_zba_zbb_zbs_zicsr`
  - `rv32ima_zce_zba_zbb_zbs_zfinx_zicsr`
  - `rv32ima_zce_zba_zbb_zbs_zfinx_zdinx_zicsr`

* RMX-500 (`-mtune=arc-v-rmx-500-series`, without hard FPU):

  - `rv32i_zce_zba_zbb_zbs_zicsr`
  - `rv32im_zce_zba_zbb_zbs_zicsr`
  - `rv32ima_zce_zba_zbb_zbs_zicsr`
  - `rv32ima_zce_zba_zbb_zbs_zfinx_zicsr`
  - `rv32ima_zce_zba_zbb_zbs_zfinx_zdinx_zicsr`

* RHX-100 (`-mtune=arc-v-rhx-100-series`, with hard FPU):

  - `rv32imac_zcb_zcmp_zba_zbb_zbs_zicsr`
  - `rv32imafc_zcb_zcmp_zba_zbb_zbs_zicsr`
  - `rv32imafd_zca_zcb_zcmp_zba_zbb_zbs_zicsr`

* RPX-100 (with `-mcmodel=medlow` and `-mcmodel=medany`):

  - `rv64imac_zcb_zba_zbb_zbs_zicsr`
  - `rv64imafdc_zcb_zba_zbb_zbs_zicsr`

Simple fallback configurations are also included:

* RMX-100 (`-mtune=arc-v-rmx-100-series`):

  - `rv32i`
  - `rv32i_zicsr`
  - `rv32im`
  - `rv32im_zicsr`
  - `rv32imc`
  - `rv32imc_zicsr`
  - `rv32imac`
  - `rv32imac_zicsr`

* RMX-500 (`-mtune=arc-v-rmx-500-series`):

  - `rv32imc`
  - `rv32imc_zicsr`
  - `rv32imac`
  - `rv32imac_zicsr`

* RHX-100 (`-mtune=arc-v-rhx-100-series`):

  - `rv32imafc`
  - `rv32imafdc`

* RPX-100:

  - `rv64imac_zicsr` (`medlow` and `medany`)
  - `rv64imafdc` (`medlow` and `medany`)

Output of `-print-multi-lib`:

```
$ riscv64-snps-elf-gcc -print-multi-lib
.;
rv32i/ilp32/rmx100;@march=rv32i@mabi=ilp32@mtune=arc-v-rmx-100-series
rv32i_zicsr/ilp32/rmx100;@march=rv32i_zicsr@mabi=ilp32@mtune=arc-v-rmx-100-series
rv32im/ilp32/rmx100;@march=rv32im@mabi=ilp32@mtune=arc-v-rmx-100-series
rv32im_zicsr/ilp32/rmx100;@march=rv32im_zicsr@mabi=ilp32@mtune=arc-v-rmx-100-series
rv32imc/ilp32/rmx100;@march=rv32imc@mabi=ilp32@mtune=arc-v-rmx-100-series
rv32imc/ilp32/rmx500;@march=rv32imc@mabi=ilp32@mtune=arc-v-rmx-500-series
rv32imc_zicsr/ilp32/rmx100;@march=rv32imc_zicsr@mabi=ilp32@mtune=arc-v-rmx-100-series
rv32imc_zicsr/ilp32/rmx500;@march=rv32imc_zicsr@mabi=ilp32@mtune=arc-v-rmx-500-series
rv32imac/ilp32/rmx100;@march=rv32imac@mabi=ilp32@mtune=arc-v-rmx-100-series
rv32imac/ilp32/rmx500;@march=rv32imac@mabi=ilp32@mtune=arc-v-rmx-500-series
rv32imac_zicsr/ilp32/rmx100;@march=rv32imac_zicsr@mabi=ilp32@mtune=arc-v-rmx-100-series
rv32imac_zicsr/ilp32/rmx500;@march=rv32imac_zicsr@mabi=ilp32@mtune=arc-v-rmx-500-series
rv32imafc/ilp32f/rhx100;@march=rv32imafc@mabi=ilp32f@mtune=arc-v-rhx-100-series
rv32imafdc/ilp32d/rhx100;@march=rv32imafdc@mabi=ilp32d@mtune=arc-v-rhx-100-series
rv64imac_zicsr/lp64;@march=rv64imac_zicsr@mabi=lp64
rv64imac_zicsr/lp64/medany;@march=rv64imac_zicsr@mabi=lp64@mcmodel=medany
rv64imafdc/lp64d;@march=rv64imafdc@mabi=lp64d
rv64imafdc/lp64d/medany;@march=rv64imafdc@mabi=lp64d@mcmodel=medany
rv32e/ilp32e/rmx100;@march=rv32e@mabi=ilp32e@mtune=arc-v-rmx-100-series
rv32e_zicsr/ilp32e/rmx100;@march=rv32e_zicsr@mabi=ilp32e@mtune=arc-v-rmx-100-series
rv32em_zce_zba_zbb_zbs_zfinx_zdinx_zicsr/ilp32e/rmx100;@march=rv32em_zce_zba_zbb_zbs_zfinx_zdinx_zicsr@mabi=ilp32e@mtune=arc-v-rmx-100-series
rv32ema_zce_zba_zbb_zbs_zfinx_zdinx_zicsr/ilp32e/rmx100;@march=rv32ema_zce_zba_zbb_zbs_zfinx_zdinx_zicsr@mabi=ilp32e@mtune=arc-v-rmx-100-series
rv32i_zce_zba_zbb_zbs_zicsr/ilp32/rmx100;@march=rv32i_zce_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=arc-v-rmx-100-series
rv32i_zce_zba_zbb_zbs_zicsr/ilp32/rmx500;@march=rv32i_zce_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=arc-v-rmx-500-series
rv32im_zce_zba_zbb_zbs_zicsr/ilp32/rmx100;@march=rv32im_zce_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=arc-v-rmx-100-series
rv32im_zce_zba_zbb_zbs_zicsr/ilp32/rmx500;@march=rv32im_zce_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=arc-v-rmx-500-series
rv32ima_zce_zba_zbb_zbs_zicsr/ilp32/rmx100;@march=rv32ima_zce_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=arc-v-rmx-100-series
rv32ima_zce_zba_zbb_zbs_zicsr/ilp32/rmx500;@march=rv32ima_zce_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=arc-v-rmx-500-series
rv32ima_zce_zba_zbb_zbs_zfinx_zicsr/ilp32/rmx100;@march=rv32ima_zce_zba_zbb_zbs_zfinx_zicsr@mabi=ilp32@mtune=arc-v-rmx-100-series
rv32ima_zce_zba_zbb_zbs_zfinx_zicsr/ilp32/rmx500;@march=rv32ima_zce_zba_zbb_zbs_zfinx_zicsr@mabi=ilp32@mtune=arc-v-rmx-500-series
rv32ima_zce_zba_zbb_zbs_zfinx_zdinx_zicsr/ilp32/rmx100;@march=rv32ima_zce_zba_zbb_zbs_zfinx_zdinx_zicsr@mabi=ilp32@mtune=arc-v-rmx-100-series
rv32ima_zce_zba_zbb_zbs_zfinx_zdinx_zicsr/ilp32/rmx500;@march=rv32ima_zce_zba_zbb_zbs_zfinx_zdinx_zicsr@mabi=ilp32@mtune=arc-v-rmx-500-series
rv32imac_zcb_zcmp_zba_zbb_zbs_zicsr/ilp32/rhx100;@march=rv32imac_zcb_zcmp_zba_zbb_zbs_zicsr@mabi=ilp32@mtune=arc-v-rhx-100-series
rv32imafc_zcb_zcmp_zba_zbb_zbs_zicsr/ilp32f/rhx100;@march=rv32imafc_zcb_zcmp_zba_zbb_zbs_zicsr@mabi=ilp32f@mtune=arc-v-rhx-100-series
rv32imafd_zca_zcb_zcmp_zba_zbb_zbs_zicsr/ilp32d/rhx100;@march=rv32imafd_zca_zcb_zcmp_zba_zbb_zbs_zicsr@mabi=ilp32d@mtune=arc-v-rhx-100-series
rv64imac_zcb_zba_zbb_zbs_zicsr/lp64;@march=rv64imac_zcb_zba_zbb_zbs_zicsr@mabi=lp64
rv64imac_zcb_zba_zbb_zbs_zicsr/lp64/medany;@march=rv64imac_zcb_zba_zbb_zbs_zicsr@mabi=lp64@mcmodel=medany
rv64imafdc_zcb_zba_zbb_zbs_zicsr/lp64d;@march=rv64imafdc_zcb_zba_zbb_zbs_zicsr@mabi=lp64d
rv64imafdc_zcb_zba_zbb_zbs_zicsr/lp64d/medany;@march=rv64imafdc_zcb_zba_zbb_zbs_zicsr@mabi=lp64d@mcmodel=medany
```

@artemiy-volkov @abrodkin Could you please review this?

**Review by artemiy-volkov — APPROVED** [review]
LGTM

**Pull request merged**